### DISCUSSION TITLE
Channel management for correlated publish confirm

### DISCUSSION THREAD

**Discussion by arn-rio** [spring-projects/spring-amqp#2637] [Q&A]
Hello,

We experienced some issues using correlated Rabbit confirm publishers. 
Rabbit server was unstable for a while. Once restored, we were unable to publish new confirmed messages to it (the max number of channel on connection was reached and the existing channels were ignored).
The solution was to restart the service (we could force close the connection : but it only worked when channel cache size wasn't set in `CachingConnectionFactory`)

We investigated and found that :
 - when a message is published but not confirmed, assigned channel is not used for an other message. (see [spring amqp documentation](https://docs.spring.io/spring-amqp/docs/3.0.8/reference/html/#template-confirms) )
 - so `CachingConnectionFactory` creates new channels, and then number of channel in connection increases.
 - Finally, there is no channel available on the connection : max number of channels reached
   when a new channelCacheSize and channelCheckoutTimeout are set in CachingConnectionFactory, we got 'No available channels' error, otherwise, 'The channelMax limit is reached'.

We call `RabbitTemplate.getUnconfirmed() `in a periodic task : the correlation data are removed, but unfortunatly, the related channel is not freed and not available for new publish.

You can observe and reproduce the issue as follows :
 - a periodic task that publishes a confirmed message on Rabbit every 20 ms 
 - a cron job that calls `RabbitTemplate.getUnconfirmed()`
 - on a local Rabbit server, block then unblock connection by updating memory watermark
```
	//block connections 
	rabbitmqctl set_vm_memory_high_watermark <low threshold>
	//unblock
	rabbitmqctl set_vm_memory_high_watermark <high threshold>
```
	
Is there a better way to manage not confirmed messages? How can we free the channels when confirmation is not received before a timeout? 

Thanks

**Comment by artembilan** [comment]
Any chances that you can share a simple project to let us to reproduce the problem?
It feels like we never close those channels which are waiting for confirms, but I'd like to be on the same page with you and be able to debug the problem.
This is the logic I'm talking about:
```
		private void returnToCache(ChannelProxy proxy) {
			if (CachingConnectionFactory.this.active && this.publisherConfirms
					&& proxy instanceof PublisherCallbackChannel) {

				this.theConnection.channelsAwaitingAcks.put(this.target, proxy);
				((PublisherCallbackChannel) proxy)
						.setAfterAckCallback(c ->
								doReturnToCache(this.theConnection.channelsAwaitingAcks.remove(c)));
			}
			else {
				doReturnToCache(proxy);
			}
		}
```

I wonder if `CachingConnectionFactory.resetConnection()` can help you somehow...

**Comment by arn-rio** [comment]
A demo : [rabbitdemo.zip](https://github.com/spring-projects/spring-amqp/files/14436587/rabbitdemo.zip)

Rabbit configuration can be updated directly in source. You should reproduce the issue by blocking and unblocking the connection on the Rabbit server while this app is running.

I did some tests by calling `CachingConnectionFactory.resetConnection()` when the `RabbitTemplate.getUnconfirmed()` result is not null. This seems to work around the connection issue (need further testing on my part).

**Reply by artembilan** [reply]
Thank for the sample!
It's a bit complicated, but I still see what is going on from the plain debugging session.
The `rabbitTemplate.convertAndSend()` does this:
```
doExecute(ChannelCallback<T> action, ConnectionFactory connectionFactory)
```
where in the end we have:
```
		finally {
			cleanUpAfterAction(channel, invokeScope, resourceHolder, connection);
		}
```
This one leads to the call in the `CachingConnectionFactory.CachedChannelInvocationHandler.invoke()` and further to the `logicalClose((ChannelProxy) proxy)`.
This one calls `returnToCache(proxy);` where we have logic like this:
```
		private void returnToCache(ChannelProxy proxy) {
			if (CachingConnectionFactory.this.active && this.publisherConfirms
					&& proxy instanceof PublisherCallbackChannel) {

				this.theConnection.channelsAwaitingAcks.put(this.target, proxy);
				((PublisherCallbackChannel) proxy)
						.setAfterAckCallback(c ->
								doReturnToCache(this.theConnection.channelsAwaitingAcks.remove(c)));
			}
			else {
				doReturnToCache(proxy);
			}
		}
```
And now that `afterAckCallback` is called only if we got some confirmation back from the broker.
According to you scenario we really never get any answer from the broker for those confirmations. Therefore channel remains opened. At least on the application side in that cache.
Or is it just a fact that channel remains in the `CachingConnectionFactory.channelList` but closed because the previous broker outage?

I wonder if there is still there some signal from Broker which we would catch to handle this kind of situation to clean channels waiting for their confirmation...

**Reply by artembilan** [reply]
Since our confirm channels are not closed normally, we fail here:
```
				if (!permits.tryAcquire(this.channelCheckoutTimeout, TimeUnit.MILLISECONDS)) {
					throw new AmqpTimeoutException("No available channels");
				}
```
And that is a bit before we would be able to know that channel is closed for some reason.

Well, at the same time we cannot know that at all because that `getCachedChannelProxy()` result is not stored anywhere in the `CachingConnectionFactory` to track its state for some reason or so...

**Reply by artembilan** [reply]
OK. Created an issue: https://github.com/spring-projects/spring-amqp/issues/2640.
Will provide the fix soon enough.

**Reply by arn-rio** [reply]
Thanks for your help and the reactivity.
I'll do more testing when the fix is available.

**Reply by artembilan** [reply]
You can give it a shot in your project with a SNAPSHOT available from the https://repo.spring.io/snapshot

**Reply by arn-rio** [reply]
Thanks, tested on my side and no more channel leaks.

In logs, I see that the channels are closed and the confirmations of expired messages are ignored.

Do you know when the next spring-amqp will be available and if it will be included in the next Springboot version?

**Reply by artembilan** [reply]
Sure! March 18th.
See our Release Calendar: https://calendar.spring.io.
Those Spring Boot versions scheduled for March 21th are going to include those Spring AMQP versions.

**Comment by arn-rio** [comment]
Hello,

We have deployed the Spring AMQP version with the fix in our production environment but we have now a new and different channel leak. The number of channels in the connection used by our confirmed publishers messages increased continuously while the service is running (around 100 new channels are created by day). 

![image](https://github.com/spring-projects/spring-amqp/assets/108938753/95f314b4-0923-4f1e-8303-ee5ec45a4f43)

This chart shows the number of channels by connection from Rabbit server side. The 2 connections that are growing are the connections that manage confirm publishes.

Before upgrading Sprinq AMQP version, on Rabbit Cluster, the number of channel opened minus the number of channel closed was 0. After the update, this value has became positive.

Do you have an idea where this could come from?

**Reply by artembilan** [reply]
That's something new.
Any chances to share with us some simple project to let us to reproduce?
Thanks

**Comment by arn-rio** [comment]
We have not reproduced it for now. The context is different from the previous leak. The number of channels grows while all servers are operationnal (and not during Rabbit server incident).

**Reply by arn-rio** [reply]
Hello, 
I tried to fix the leak by updating the connection settings. When `ChannelCacheSize` and `ChannelCheckoutTimeout` are initialized, there have been no unpublished messages and the number of chanels used is lower than when these parameters are unset.

This seems to be a good fix but I saw in my tests that after a Rabbit connection loss, the  `ChannelCacheSize` limit is often ignored and the number of open channels exceeds `ChannelCacheSize` . 

You can reproduce this case with the following scenario  : 
- run Rabbit server
- run demo (after updating rabbit connection parameters)
- check that the number of channel is lower or equal than ChannelCacheSize
- close connection from Rabbit server side
- check again the number of channel in the connection (in my case it's always fluctuating and can be greater than ChannelCacheSize)

Demo has been updated since last time : [rabbitdemo.zip](https://github.com/spring-projects/spring-amqp/files/15306949/rabbitdemo.zip)

Thanks

**Reply by artembilan** [reply]
Sorry, I'm not fully sure what I need to change in your demo to reproduce "channels exceeds ChannelCacheSize".
Do you mean to comment out this part of your code:
```
        connectionFactory.setChannelCacheSize(10);
        connectionFactory.setChannelCheckoutTimeout(5000);
```
?

**Reply by arn-rio** [reply]
No, sorry, I didn't mean changes in the connection configuration, I mean changes in the Rabbit server config (server address, exchange, ...) according to your test environment. I also forgot to update pom in demo, update  spring-boot-starter-amqp  to 3.1.11 for and spring-rabbit to 3.0.13. 

To reproduce the issue, I just need to run the demo and then close the connection while it is running after the messages have been published.

![image](https://github.com/spring-projects/spring-amqp/assets/108938753/8c17a01d-80a3-455b-bda8-77d32b0ba455)

**Reply by artembilan** [reply]
OK. Will try. 
Is there a way to reproduce it in not so stressful manner:
```
 @Scheduled(fixedRate = 50)
```
?

I mean that's too many, too often messages to publish.
Hard to follow logs and debug session.

**Reply by arn-rio** [reply]
The probability to reproduce the problem decreases if we increase the fixedRate value.

During the issue, publish confirmed client creates and removes many channels (the channel id in the connection is always greater)

**Reply by artembilan** [reply]
```
2024-05-16T13:18:18.040-04:00 DEBUG 128412 --- [ectionFactory11] o.s.a.r.c.CachingConnectionFactory       : Closing cached Channel: PublisherCallbackChannelImpl: AMQChannel(amqp://guest@127.0.0.1:5672/,417)
2024-05-16T13:18:18.040-04:00 DEBUG 128412 --- [ectionFactory22] o.s.a.r.c.PublisherCallbackChannelImpl   : PendingConfirms cleared
2024-05-16T13:18:18.040-04:00 DEBUG 128412 --- [ectionFactory23] o.s.a.r.c.PublisherCallbackChannelImpl   : PendingConfirms cleared
2024-05-16T13:18:18.040-04:00 DEBUG 128412 --- [ectionFactory28] o.s.a.r.c.CachingConnectionFactory       : Released permit for 'Proxy@3f163da0 Shared Rabbit Connection: SimpleConnection@683cba3f [delegate=amqp://guest@127.0.0.1:5672/, localPort=60888]', remaining: 415
```
So, apparently there is unconditional `permits.release()` which increases the size of that `Semaphore`  causing this leak after re-connection.

Investigating further...

**Reply by arn-rio** [reply]
It looks like what I saw.

**Reply by artembilan** [reply]
I made some fix: https://github.com/spring-projects/spring-amqp/issues/2715.
It is not perfect, but at least it mitigates a leak a little bit.
I still cannot determine ho those channel may come back to the cache if we have this condition:
```
this.theConnection.channelListLock.lock();
				try {
					// Allow for multiple close calls...
					if (CachingConnectionFactory.this.active) {
						cacheOrClose(proxy);
					}
...

		private void cacheOrClose(ChannelProxy proxy) {
			boolean alreadyCached = this.channelList.contains(proxy);
			if (this.channelList.size() >= getChannelCacheSize() && !alreadyCached) {
				if (logger.isTraceEnabled()) {
					logger.trace("Cache limit reached: " + this.target);
				}
				try {
					physicalClose();
				}
				catch (@SuppressWarnings(UNUSED) Exception e) {
				}
			}
			else if (!alreadyCached) {
				if (logger.isTraceEnabled()) {
					logger.trace("Returning cached Channel: " + this.target);
				}
				this.channelList.addLast(proxy);
				releasePermitIfNecessary();
				setHighWaterMark();
			}
		}
```

Will look into that later since we have to release today at least some fix.

Thank you for your feedback!

**Reply by arn-rio** [reply]
Hello, I was on vacation last week. I've tested spring-rabbit version 3.0.14 this morning and noticed the improvement. There's still a minor leak when a new connection is established, but it disappears after a few seconds.  The number of channels remains below the threshold.

Thanks for the fix.
